src/block.ts

```diff
@@ -0,0 +1,269 @@
+   * Ethereum wire protocol. If `true`, a tuple with the raw data of the header, the txs and the
+   * uncle headers is returned.
+   */
+  serialize(): Buffer
+  serialize(rlpEncode: true): Buffer
+  serialize(rlpEncode: false): [Buffer[], Buffer[], Buffer[]]
+  serialize(rlpEncode = true) {
+    const raw = [
+      this.header.raw,
+      this.transactions.map(tx => tx.raw),
+      this.uncleHeaders.map(uh => uh.raw),
+    ]
+
+    return rlpEncode ? rlp.encode(raw) : raw
+  }
+
+  /**
+   * Generate transaction trie. The tx trie must be generated before the transaction trie can
+   * be validated with `validateTransactionTrie`
+   */
+  async genTxTrie(): Promise<void> {
+    for (let i = 0; i < this.transactions.length; i++) {
+      const tx = this.transactions[i]
+      await this._putTxInTrie(i, tx)
+    }
+  }
+
+  /**
+   * Validates the transaction trie
+   */
+  validateTransactionsTrie(): boolean {
+    const txT = this.header.transactionsTrie.toString('hex')
+    if (this.transactions.length) {
+      return txT === this.txTrie.root.toString('hex')
+    } else {
+      return txT === ethUtil.KECCAK256_RLP.toString('hex')
+    }
+  }
+
+  /**
+   * Validates the transactions
+   *
+   * @param stringError - If `true`, a string with the indices of the invalid txs is returned.
+   */
+  validateTransactions(): boolean
+  validateTransactions(stringError: false): boolean
+  validateTransactions(stringError: true): string
+  validateTransactions(stringError = false) {
+    const errors: string[] = []
+
+    this.transactions.forEach(function(tx, i) {
+      const error = tx.validate(true)
+      if (error) {
+        errors.push(`${error} at tx ${i}`)
+      }
+    })
+
+    if (!stringError) {
+      return errors.length === 0
+    }
+
+    return errors.join(' ')
+  }
+
+  /**
+   * Validates the entire block, throwing if invalid.
+   *
+   * @param blockChain - the blockchain that this block wants to be part of
+   */
+  async validate(blockChain: Blockchain): Promise<void> {
+    await Promise.all([
+      this.validateUncles(blockChain),
+      this.genTxTrie(),
+      this.header.validate(blockChain),
+    ])
+
+    if (!this.validateTransactionsTrie()) {
+      throw new Error('invalid transaction trie')
+    }
+
+    const txErrors = this.validateTransactions(true)
+    if (txErrors !== '') {
+      throw new Error(txErrors)
+    }
+
+    if (!this.validateUnclesHash()) {
+      throw new Error('invalid uncle hash')
+    }
+  }
+
+  /**
+   * Validates the uncle's hash
+   */
+  validateUnclesHash(): boolean {
+    const raw = rlp.encode(this.uncleHeaders.map(uh => uh.raw))
+
+    return ethUtil.keccak256(raw).toString('hex') === this.header.uncleHash.toString('hex')
+  }
+
+  /**
+   * Validates the uncles that are in the block, if any. This method throws if they are invalid.
+   *
+   * @param blockChain - the blockchain that this block wants to be part of
+   */
+  async validateUncles(blockchain: Blockchain) {
```

Comment:
Return value is missing here.

Reply:
Adding it made me realize that the return type was super bizarre. I simplified this function.